File: 1/1.go
package main

import (
	"aoc24/utils"
	"math"
	"sort"
	"strconv"
	"strings"
)

func main() {
	fileContent := utils.ReadFileReturnArray("./inputs/day1-input.txt")

	var left []int
	var right []int

	for _, line := range fileContent {
		split := strings.Fields(line)

		leftint, _ := strconv.Atoi(split[0])
		rightint, _ := strconv.Atoi(split[1])

		left = append(left, leftint)
		right = append(right, rightint)
	}

	sort.Ints(left)
	sort.Ints(right)

	sum := 0

	for i, leftItem := range left {
		sum += int(math.Abs(float64(leftItem - right[i])))
	}

	utils.Solution(sum, "a")

	rightMap := make(map[int]int)

	for _, rightItem := range right {
		if _, exists := rightMap[rightItem]; !exists {
			rightMap[rightItem] = 1
		} else {
			rightMap[rightItem] = rightMap[rightItem] + 1
		}
	}

	sum2 := 0
	for _, leftItem := range left {
		if val, exists := rightMap[leftItem]; exists {
			sum2 += leftItem * val
		}
	}

	utils.Solution(sum2, "b")
}

File: 1/1.js
const {readFileToArray, debug, solution} = require('../utils/util')

const input = './inputs/day1-input.txt'


async function main() {
    const data = await readFileToArray(input)

    debug('input length: ', data.length)

    let left = [], right = []

    for (eachRow of data) {
        const split = eachRow.split("   ")

        if (split.length !== 2) {
            debug('split length is not 2', split)
            continue
        }

        if (!isNaN(Number(split[0]))) {left.push(Number(split[0]))}
        if (!isNaN(Number(split[1]))) {right.push(Number(split[1]))}
    }

    debug(left.length, right.length)

    left = left.sort()
    right = right.sort()


    const sumDistance = left
        .flatMap((v, i) => Math.abs(v - right[i]))


    solution('a', sumDistance.reduce((acc, curr) => acc + curr, 0))

    const rightMap = {}
    right.forEach((v) => {
        if (!rightMap[v]) {
            rightMap[v] = 1
        } else {
            rightMap[v] = rightMap[v] + 1
        }
    })


    solution('b', left.reduce((acc, curr) => {
        if (!rightMap[curr]) {
            return acc
        } else {
            return acc + (curr * rightMap[curr])
        }
    }, 0))
}

main().catch(console.error)